codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -328,15 +334,37 @@ class JsonParserGenerator(
         } else if (target.isDoubleShape) {
             rustTemplate("#{expect_number_or_null}(tokens.next())?.map(|v| v.to_f64_lossy())", *codegenScope)
         } else {
-            rustTemplate(
-                """
-                #{expect_number_or_null}(tokens.next())?
-                    .map(#{NumberType}::try_from)
-                    .transpose()?
-                """,
-                "NumberType" to symbolProvider.toSymbol(target),
-                *codegenScope,
-            )
+            val numberSymbol = symbolProvider.toSymbol(target)
+            if (numberSymbol.rustType() is RustType.Opaque) {
+                val precision = when (target) {
+                    is LongShape -> 64
+                    is IntegerShape -> 32
+                    is ShortShape -> 16
+                    is ByteShape -> 8
+                    else -> {
+                        throw IllegalStateException("unreachable")
+                    }
+                }
+                rustTemplate(
+                    """
+                    #{expect_number_or_null}(tokens.next())?
+                        .map(#{NumberPrimitive}::try_from)
+                        .transpose()?
+                    """,
+                    "NumberPrimitive" to RustType.Integer(precision),
+                    *codegenScope,
+                )
+            } else {
```

Comment:
`returnSymbolToParse` should help you in coalescing both arms, giving you the unconstrained symbol for both cases.

Reply:
Done in https://github.com/awslabs/smithy-rs/pull/2005/commits/a959004accbdde4e577482c8b47a347eb1acffc9